File: DynamicProgramming/WordBreak/break.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution1 {
public:
    unordered_map<string,bool> dp;
    bool solve(string s, vector<string>& wordDict) {
        if(dp.count(s))return dp[s];

        if(s.empty())
            return true;

        for(string word : wordDict) {

            // Check whether word matches at the beginning
            if(s.substr(0, word.size()) == word) {

                // Remove matched word and solve remaining string
                string remaining = s.substr(word.size());

                if(solve(remaining, wordDict))
                    return dp[s] = true;
            }
        }

        return dp[s] = false;
    }

    bool wordBreak(string s, vector<string>& wordDict) {
        return solve(s, wordDict);
    }
};

class Solution2 {
public:
    vector<int> dp;

    bool doDP(int i, string& s, unordered_set<string>& dict,
              vector<int>& lengths) {

        if(i == s.size())
            return true;

        if(dp[i] != -1)
            return dp[i];

        for(int len : lengths) {

            if(i + len > s.size())
                continue;

            string word = s.substr(i, len);

            if(dict.count(word)) {
                if(doDP(i + len, s, dict, lengths))
                    return dp[i] = 1;
            }
        }

        return dp[i] = 0;
    }

    bool wordBreak(string &s, vector<string> &dictionary) {

        unordered_set<string> dict;
        unordered_set<int> lengthSet;

        for(string& word : dictionary) {
            dict.insert(word);
            lengthSet.insert(word.size());
        }

        vector<int> lengths(lengthSet.begin(), lengthSet.end());

        dp.assign(s.size(), -1);

        return doDP(0, s, dict, lengths);
    }
};

File: DynamicProgramming/WordBreak/break.py
class Solution1:
    def __init__(self):
        self.dp = {}

    def solve(self, s, wordDict):
        if s in self.dp:
            return self.dp[s]

        if not s:
            return True

        for word in wordDict:
            if s[:len(word)] == word:

                remaining = s[len(word):]

                if self.solve(remaining, wordDict):
                    self.dp[s] = True
                    return True

        self.dp[s] = False
        return False

    def wordBreak(self, s, wordDict):
        return self.solve(s, wordDict)


class Solution2:
    def __init__(self):
        self.dp = []

    def doDP(self, i, s, dictionary, lengths):
        if i == len(s):
            return True

        if self.dp[i] != -1:
            return self.dp[i]

        for length in lengths:

            if i + length > len(s):
                continue

            word = s[i:i + length]

            if word in dictionary:
                if self.doDP(i + length, s, dictionary, lengths):
                    self.dp[i] = 1
                    return True

        self.dp[i] = 0
        return False

    def wordBreak(self, s, dictionary):
        self.dp = [-1] * len(s)

        dictionary = set(dictionary)
        lengths = set(len(word) for word in dictionary)

        return self.doDP(0, s, dictionary, lengths)